design/design.md

```diff
@@ -0,0 +1,223 @@
+to `io.xpring.v2` results in objects with namespace `io::xpring::v2`. In the event
+that a handler does not change between versions (which implies the objects are
+the same in all but namespace), the old handler can be reused by templating the
+handler signature like so:
+
+```
+template<class T, class R>
+R doFee(RPC::ContextGeneric<T>& context);
+```
+## Use Case
+The GRPC service will be used in a request response format (as opposed to streaming).
+The client will issue a request, and then receive a response.
+
+An application developer can quickly write client code in their language of
+choice using the `.proto` files and the code generation tools of gRPC.
+
+## Interface
+The interface to the GRPC service is defined by the .proto files. The `.proto` files
+define the structure of the messages, the methods available, their inputs and outputs, and the service itself.
+GRPC, with the help of protobuf, autogenerates code based on these `.proto` files,
+for both client and server. GRPC has bindings in many common programming languages
+([link](https://grpc.io/docs/reference/)).
+A client of the GRPC service we are building would use the exact same `.proto`
+files as the server, to generate the client code. The generated code (both
+client and server), handles serialization, deserialization and I/O. Protobuf
+allows one to define messages with named and typed fields, and allows for user
+defined types. Client and server code can then manipulate the generated objects using the
+message names, field names and types defined in the `.proto` file. For example,
+using the below example `.proto` file:
+```
+service XRPLedgerAPI {
+  // Get account info for an account on the XRP Ledger.
+  rpc GetAccountInfo (GetAccountInfoRequest) returns (AccountInfo);
+
+  // Get the fee for a transaction on the XRP Ledger.
+  rpc GetFee (GetFeeRequest) returns (Fee);
+  message GetAccountInfoRequest {
+    // The address to get info about.
+    string address = 1;
+  }
+
+  message AccountInfo {
+    XRPAmount balance = 1;
+  
+    uint64 sequence = 2;
+  
+  }
+  message XRPAmount {
+    // A numeric string representing the number of drops of XRP.
+    string drops = 1;
+  }
+
+
+}
+```
+A client can do things like:
+```
+GetAccountInfoRequest request;
+request.set_address("rLkMJhSVwhmummLjJPVrwQRZZYiYQhVQ1A");
+//Send the request, receive populated response
+uint64_t sequence = response.sequence();
+std::string drops = response.balance.drops();
+```
+
+For more information about protocol buffers see
+[here](https://developers.google.com/protocol-buffers/docs/overview).
+For more information about grpc and protocol buffers, see
+[here](https://grpc.io/docs/guides/)
+
+## Sequence Diagrams?
+Below is a sequence diagram of handling two grpc requests in parallel. In the diagram, the responses are sent in the order the requests were received; this is not a rule, and responses could be sent in a different order than the requests were recieved.
```

Comment:
2 sequence diagram I propose to have
1. Life time of a CallData 
2. How is one call being executed internally from start to finish.